src/rust/engine/fs/store/src/lib.rs

```diff
@@ -1377,20 +1368,23 @@ impl Store {
     perms: Permissions,
     is_executable: bool,
     can_be_immutable: bool,
-    immutable_inputs: Option<&ImmutableInputs>,
   ) -> Result<(), StoreError> {
-    if can_be_immutable && digest.size_bytes > IMMUTABLE_FILE_SIZE_LIMIT {
-      let dest_path = immutable_inputs
-        .unwrap()
-        .path_for_file(digest, is_executable)
-        .await?;
-      self
-        .materialize_hardlink(destination, dest_path.to_str().unwrap().to_string())
-        .await
+    let hardlink_dest = if can_be_immutable {
+      self.local.load_from_fs(digest).await?
     } else {
-      self
-        .materialize_file(destination, digest, perms, is_executable)
-        .await
+      None
+    };
+    match hardlink_dest {
+      Some(path) => {
+        self
+          .materialize_hardlink(destination, path.to_str().unwrap().to_string())
+          .await
+      }
+      None => {
+        self
+          .materialize_file(destination, digest, perms, is_executable)
+          .await
```

Comment:
Does `materialize_file` need an update as well? It's going to use `load_file_bytes_with` still, right? But it should be able to use `copy` for large files now.

Reply:
No this one truly materializes a file. The hardlink source is immutable to guard against shenanigans and we only ever call this if the file _must_ be mutable.